text/0016-generalized-entity-store.md

```diff
@@ -0,0 +1,113 @@
+# Generalized Entity Store
+
+## Related issues and PRs
+
+- Implementation PR(s): https://github.com/cedar-policy/cedar/pull/171, which is a pull request from https://github.com/prakol16/cedar
```

Comment:
I think we might want to extend what you have so there are three functions in the interface:
- `get_entity`
-  `eval_in`
- `get_record`

`get_entity` would be essentially what you have. `eval_in` would replace the need for an `ancestors` attribute on the entity you return. `get_record` would allow you to return an entity `E` without filling out all of the data in a (potentially large) record attribute of `E`.

It might be worth adding `to_be_loaded_Entity/Record` variants to signal that they should be loaded from the database, otherwise attempting to access them would be an error.

My instinct is that we should use a `visitor` so that calling `get_attr` for `E.entity_A` could do more than just return `entity_A` (e.g., start collecting `entity_B` that I know will be requested next). It's not clear to me if we would then keep the interface as {`get_entity`, `eval_in`, `get_record`} or if the visitor should block and update the requested entity data in the background before returning and the interface would just be calling the visitor.

Reply:
Note that we can add `get_entity` and `eval_in` now without adding `to_be_loaded_Entity/Record` and without the visitor and we can do the rest later. (Unless we go with just the visitor in which case we'd need some breaking changes, but I think they'd be ok.)

Reply:
I'm a little confused about how you're picturing `get_record`. If the type signature is something like `get_record(uid: EntityUID, attr: SmolStr) -> Result<PartialValue>`, which gets the attribute `attr` from entity with uid `uid`, then I think a better name would be `get_attr`.

Note that in addition to `get_attr` we would also need `has_attr` (to implement `has`). Finally, with these two, I'm not sure if we need `get_entity` at all anymore -- it could probably be eliminated or replaced with `exists_entity(uid) -> bool` which returns whether the entity with the given uid exists (apparently sometimes for the spec we need to check if an entity exists without ever actually having to fetch it).